javaslang/src/main/java/javaslang/collection/IndexedSeq.java

```diff
@@ -221,16 +223,68 @@ default int lastIndexWhere(Predicate<? super T> predicate, int end) {
     @Override
     default Iterator<T> reverseIterator() {
         return new AbstractIterator<T>() {
-            private int i = IndexedSeq.this.length();
+            final ListIterator<T> delegate = IndexedSeq.this.listIterator(IndexedSeq.this.length());
 
             @Override
             public boolean hasNext() {
-                return i > 0;
+                return delegate.hasPrevious();
             }
 
             @Override
             public T getNext() {
-                return IndexedSeq.this.get(--i);
+                return delegate.previous();
+            }
+        };
+    }
+
+    default ListIterator<T> listIterator(int index) {
+        return new ListIterator<T>() {
+            private final IndexedSeq<T> delegate = IndexedSeq.this;
+            private int i = index;
+
+            @Override
+            public boolean hasNext() {
+                return i < delegate.length();
+            }
+
+            @Override
+            public T next() {
+                return delegate.get(i++);
+            }
+
+            @Override
+            public int nextIndex() {
+                return i;
+            }
+
+            @Override
+            public boolean hasPrevious() {
+                return i > 0;
+            }
+
+            @Override
+            public T previous() {
+                return delegate.get(--i);
+            }
+
+            @Override
+            public int previousIndex() {
+                return i - 1;
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
```

Comment:
We should be careful with implementing logic that do break the applications of our users. I know that it is not possible in each case but Javaslang should not throw any 'UnsupportedOperationException'. I will update the CONTRIBUTION guidelines accordingly!